### ISSUE TITLE
drag heuristics can make it hard to place items (ex: web demo)

### ISSUE THREAD

**Issue by YuriGal** [gridstack/gridstack.js#1094]
Is there a way to alter behavior when one widget is dragged on top of other?  In the demo below I am trying to insert widget "7" between "9" and "2". If I drag it directly into the gap - it assumes I want insert "7" above "2" and pushes "2" down even thought I drag "7" thru the lower portion of "2". Only if I drag "7" all the way around and over "9" directly - it moves "9" down.

<img src="https://i.imgur.com/pxjSzX0.gif">

Is it possible to keep the widget that is being dragged over in place to a certain point (say when widget being dragged is at over 50% of its height) ?

**Comment by adumesny** [comment]
I agree, the heuristics could use some tweaking, and be more like over the second half vs the first half - it's also not clear if it uses the mouse position or just the box outline (or something in between).

**Comment by ataft** [comment]
I'm having a similar issue when trying to drag to the bottom. As you can see in the demo below, I have to drag the widget about 90% over the bottom widgets in order for it to move:

![gridstack_issue](https://user-images.githubusercontent.com/11670864/73802061-40aa4a80-4771-11ea-9318-8515f7c29f87.gif)

**Comment by Prid13** [comment]
Any update on this? I'm looking into gridstack.js and it's quite promising, but the weird behavior when dragging and dropping is counter intuitive to me.

Like @ataft demonstrated, dragging an element **below** another requires about 90% of height to be below the target in order to work.

Any fix in the works, or options to tweak this as per personal preference? :)

**Comment by adumesny** [comment]
on my very long todo list... but first get 2.0 out, then likely focus on this.

**Comment by ranjeettamil** [comment]
@adumesny  Waiting for features from very long time, 
Please add this feature in upcoming release.

**Comment by adumesny** [comment]
@ranjeettamil are you willing to pay for it ? there is a donate button (I sure didn't see your $$ contributions) and very few people bother contributing or helping with this project, yet expect things to be magically done for them for free... and silence...

**Comment by Prid13** [comment]
@adumesny: I'm sure he didn't mean it like that. How big of a donation are we talking about to get this feature in motion asap? I'm willing to chip in and @ranjeettamil maybe you are too? :)

**Comment by adumesny** [comment]
@Prid13 only @ataft has donated $50 so far which is very much appreciated. heck, if most people that use gridstack would even give 'beer money', someone could possibly make this a fulltime gig... never heard from @ranjeettamil again.
If a few people are willing to back this one, I'll take a look. I'm pretty sure this is not an easy one at all... been tried before (a reason #141 has been sitting for 5.5 years!)

**Comment by ranjeettamil** [comment]
@adumesny I am developer and working for client, I already convey and shared the link to them.

**Comment by Prid13** [comment]
@adumesny I would also only be able to donate a $50 at max, unfortunately.

I'm thinking of using this in a Chrome extension I'm working on that I'm also hoping to monetize. If I ever make money off of it, I'll be sure to donate a generous amount towards the development of gridstack.js :smile:

**Comment by adumesny** [comment]
I may have some time during the holidays to work on this... I think I can get the **instant swapping of items of the same size** (rather than require 90% down) first - that would also solve the use case where the grid is full and you can't move anything in any directions, and having a 1 row of items and not being able to move sideways.

the two example with very different sized items and involving moving more than one collision is much more complicated and I don't fully understand the current collision check yet. And is not based on % coverage. 

get that donation going!

**Comment by pawel-k-kazmierczak** [comment]
Hi @adumesny, here is research about how drag and drop could works:

https://www.nngroup.com/articles/drag-drop/

I hope it will be useful :)

**Comment by adumesny** [comment]
@pawel-k-kazmierczak thanks, I'll take a read. The problem isn't knowing how/when it should swap (I have a pretty UX instinct having designed many programs for Macromedia/Adobe used my millions, and now Vmware) but rather implementing it in the current framework.

**Comment by pawel-k-kazmierczak** [comment]
Thank you @adumesny for explanation!

'this does not work if the dragged items is much larger than the other item' I agree- I think that for this case it could be 'center of smaller object'- not matter if dragged or not. Maybe research by Nielsen Norman Group was made on small elements only?

Maybe the issue is lack of symmetry in moving bottom=>up (today it works a bit too fast) and up=>bottom (it works to slow). Symmetry could enhance experience for users as it would result in one behavior in each direction.

This works quite fine (but only for react):

https://strml.github.io/react-grid-layout/examples/1-basic.html

**Comment by adumesny** [comment]
actually it's not ideal either as it requires you yo drag the item way past the 50% coverage as well by quite a bit

https://user-images.githubusercontent.com/3496166/107386958-38ce6600-6aa9-11eb-8308-ac9a932c7553.mp4

and is not symmetrical either (looks like mid point going down, but above mid point by entire dragged item going up (instead of using top edge).

https://user-images.githubusercontent.com/3496166/107387234-792de400-6aa9-11eb-8911-396482445452.mp4

but still more usable...

**Comment by Prid13** [comment]
@adumesny that's unforunate. Take a look at this drag-n-drop library, it has _really_ good heuristics imo: http://hootsuite.github.io/grid/

That would be the goal, it's very intuitive to use. I'd just love more functionality with it, like panels being resizable (drag corner to resize) and other options :)

**Comment by adumesny** [comment]
@Prid13 I don't think that's intuitive at all... but to each his own. sometime moves vertical, sometimes push to the right. But hey you are welcome to use it instead...

https://user-images.githubusercontent.com/3496166/107536071-3be35800-6b76-11eb-97d2-f9e5611e94d9.mp4

**Comment by Prid13** [comment]
@adumesny ah, that's problematic, I agree. It could certainly use improvements, but so far out of all the drag-n-drop libraries I've tried, that one feels the most natural in some areas (not perfect, though).

I'd love to use it, but I feel it's lacking some options gridstack.js has, which is why I'm patiently waiting for this to be resolved in the best way possible :) Also you said you'd worked for Adobe/Macromedia?? That's super impressive and adds a lot to this project's credibility

**Comment by adumesny** [comment]
**and donations are pouring in!!!!** (I wish)

this fixes the SECOND part on this bug (moving up/down is now symmetrical) - read commit.
TODO part II: handle mid point of dragged over items rather 50% of row/column
and check for the most covered when multiple items collide.

default `float:false` (top gravity)

https://user-images.githubusercontent.com/3496166/107982087-4340a200-6f78-11eb-8b39-5a9b448a610d.mp4


`float:true` case

https://user-images.githubusercontent.com/3496166/107982494-045f1c00-6f79-11eb-9d84-14876431287a.mp4


`maxRow:3` for full grid, even if float=true

https://user-images.githubusercontent.com/3496166/107982687-5c961e00-6f79-11eb-98fd-a114196b3381.mp4

**Comment by pawel-k-kazmierczak** [comment]
Thank you @adumesny it works much more better!

I runed it locally and I saw that for hiegher elements (at least 3 rows) it dancing a bit:

https://user-images.githubusercontent.com/13885829/108404586-55446e00-7220-11eb-89f5-2f22221626ef.mp4

I hope it will be useful :)

**Comment by adumesny** [comment]
and for those wondering, I'm still working on part 2 where we only swap when passing mid point (which I have working in a branch) and won't release until that is done, and drag between grid has also been redone... it will be v4

**Comment by adumesny** [comment]
showing 50% coverage (sized 2 instead of 1 items)
https://user-images.githubusercontent.com/3496166/108586804-95c9e780-7305-11eb-8733-be7dec94284f.mp4

web demo (still some issues left, especially placing below)
https://user-images.githubusercontent.com/3496166/108586908-0a048b00-7306-11eb-97f1-8ae7adda33fe.mp4

**this stuff is HARD**!!

**Comment by adumesny** [comment]
more fixes to make this example better

https://user-images.githubusercontent.com/3496166/108612321-716d1a00-739c-11eb-8d71-70c8d69f7405.mp4

**Comment by pawel-k-kazmierczak** [comment]
Thank you @adumesny it bring great user experience!

I found two cases where it could be unintuitive now.

https://user-images.githubusercontent.com/13885829/109331982-cbb72080-785d-11eb-8e42-13c8736a3d4a.mp4

https://user-images.githubusercontent.com/13885829/109331994-d1146b00-785d-11eb-91de-48fa31ec4eea.mp4

I hope it will be useful :)

**Comment by adumesny** [comment]
whoa, first one is funky! can you reproduce with the latest code ? I'll re-open.

second: right now it only swap if they cross 50% horizontally... that heuristic still needs tweeking as #1628 was surprising to have to use the lower coverage of the two... unless the issue was from dargging out and back in (might be more issues left there)

**Comment by adumesny** [comment]
fixed first case... checking on second.

https://user-images.githubusercontent.com/3496166/109376459-6e30cd00-7879-11eb-90dc-5fa02927d09c.mp4

**Comment by pawel-k-kazmierczak** [comment]
I just tested it. For first case there is one edge case when item is taken away from container:

https://user-images.githubusercontent.com/13885829/109384051-a15f7480-78ea-11eb-970b-32d5d4c5ff4d.mp4

Thank you!

**Comment by adumesny** [comment]
I'm not seeing an issue with ^^^ you are dragging it outside the grid so it gets temporarily removed, but then re-inserted.

**Comment by pawel-k-kazmierczak** [comment]
> I'm not seeing an issue with ^^^ you are dragging it outside the grid so it gets temporarily removed, but then re-inserted.

Yes, but after that '0' and '1' are swapped. Maybe I don't explained it correctly..
In last time '2' and '3' are swapped.

**Comment by adumesny** [comment]
ok, I see it now. yep still an issue. this is getting incredibly hard and I need to take a break for all of this...

**Comment by adumesny** [comment]
think I finally nailed it after spending the weekend (again!) on it fine tuning it and advanced cases I had not tried before ('load 3' is particularly tricky). Will be in v4

At some point I can further tune the 50% overage (so either shape will trigger) and if before or after the shape (tiny cell moving over the large case example could be improved.

https://user-images.githubusercontent.com/3496166/111097805-72a6e800-84ff-11eb-9ee1-f177c9bc6ec6.mp4

**Comment by pawel-k-kazmierczak** [comment]
Hi,
I'm not sure wheather it is related to this issue but I notice unexped behaviour while drag and drop:

https://user-images.githubusercontent.com/13885829/115019842-04689300-9eba-11eb-9231-45e86251f19f.mp4

-for longer widget it works
-for shorter is not possible to drop it from top to bottom.

thanks

**Comment by adumesny** [comment]
^^^ that's #219 (in your second example we need to add extra rows) which I've fixed in general **as long as the cursor stays** in the grid (we don't do shape leaving event, but cursor driven), but for some reason that page example isn't growing the grid. I noticed that too but have not looked at it. May want to post your vid to the other bug (and see mine similar that do work). AGain that webpage doesn't work as the other....
